### ISSUE TITLE
Connection refused from istiod

### ISSUE THREAD

**Issue by eddiewang** [squat/kilo#60]
After setting up kilo on k3s, i noticed that although Kilo is running, istio-ingressgateway pod can't seem to reach the istiod pod. in the logs you'll notice the the connection is refused.

I'm using coredns with kilo's CNI (no flannel).
Any advice would be greatly appreciated!

```
2020-06-12T23:29:50.578577Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:54] Unable to establish new stream
2020-06-12T23:29:50.989714Z	warn	cache	resource:default request:54d7ab7d-1db1-4fb5-ae9b-1b33c85912ee CSR failed with error: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup istiod.istio-system.svc on 10.43.0.10:53: read udp 10.42.2.9:51000->10.43.0.10:53: read: connection refused", retry in 6400 millisec
2020-06-12T23:29:50.991787Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamSecrets gRPC config stream closed: 14, connection error: desc = "transport: Error while dialing dial tcp: lookup istiod.istio-system.svc on 10.43.0.10:53: read udp 10.42.2.9:51000->10.43.0.10:53: read: connection refused"
2020-06-12T23:29:50.990179Z	error	citadelclient	Failed to create certificate: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup istiod.istio-system.svc on 10.43.0.10:53: read udp 10.42.2.9:51000->10.43.0.10:53: read: connection refused"
2020-06-12T23:29:50.990266Z	error	cache	resource:default request:54d7ab7d-1db1-4fb5-ae9b-1b33c85912ee CSR retrial timed out: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup istiod.istio-system.svc on 10.43.0.10:53: read udp 10.42.2.9:51000->10.43.0.10:53: read: connection refused"
2020-06-12T23:29:50.990327Z	error	cache	resource:default failed to generate secret for proxy: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup istiod.istio-system.svc on 10.43.0.10:53: read udp 10.42.2.9:51000->10.43.0.10:53: read: connection refused"
2020-06-12T23:29:50.990366Z	error	sds	resource:default Close connection. Failed to get secret for proxy "router~10.42.2.9~istio-ingressgateway-74d4d8d459-wlt8p.istio-system~istio-system.svc.cluster.local" from secret cache: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp: lookup istiod.istio-system.svc on 10.43.0.10:53: read udp 10.42.2.9:51000->10.43.0.10:53: read: connection refused"
```

**Comment by squat** [comment]
Hi @eddiewang, thanks for filing this :)
I'm glad to hear that you're trying Istio with Kilo. They should be complimentary technologies and I would love to have a doc describing how to use them together.

From the looks of it, it's not really an Istio problem but more fundamental; it seems that in-cluster DNS is broken. This may mean that all pod-pod and pod-service traffic is down. That's said, I've never used Istio myself, so I'm a little unfamiliar with the problem you are presenting. Can you share a bit more information about your topology? In other words, is it one cluster or many connected clusters? Is the cluster multi-cloud/multi-region? Or are you only using Kilo for VPN access to cluster IPs? What cloud provider are you running in? Does networking in the cluster work in general and is it only Istio that is broken? Or is all networking broken? Finally, where are these Istio Pods running? Are they in the same cluster, cloud, and subnet?

I think the answers to these questions should be enough to get us started in debugging the issues you see.

**Comment by squat** [comment]
PS is this the issue you are describing in https://github.com/squat/kilo/issues/26#issuecomment-643460169?

**Comment by eddiewang** [comment]
Appreciate the quick reply as always. I was actually able to debug the issue. I had a cluster of k3s servers setup with kilo is full mesh granularity. The servers were behind a floating ip. For some reason, kilo interpreted that all three machines had the same external endpoint (the floating ip). 
I was able to discover this by ssh'ing into the machine and checking the `wg` status.

I was able to manually annotate with the `kilo.squat.ai/force-endpoint=my.ip.goes.here:51820`. This kick-started the connection :)

The issue i'm describing up there is slightly related. Using `k3sup` as the cluster build tool, there was an error with the `kubeconfig` not existing on agent nodes (in the path `/etc/rancher/k3s/k3s.yaml`). 

I was able to fix it by copying over the .yaml from the master cluster to the agent nodes. However, I'm not sure what I did with my latest setup, but I was able to avoid manually copying the `k2s.yaml` and instead simply by removing the kubeconfig flag, kilo was able to operate (not sure how that works, assuming it uses incluster dns to resolve).

**Comment by eddiewang** [comment]
Another question I had was the importance of using the internal ip. Atm, my cluster is setup in a way in such that both the internal and external ip is the external ip. So if you typed `kubectl get nodes -o wide` but the `Internal-Ip` and `External-Ip` represent the publically reachable ipv4 address.

Will this cause issues in the future, for not properly setting up the internal ip?

**Comment by squat** [comment]
Wow it sounds like you made a lot of progress!

So, you can always run Kilo without mounting and specifying the Kubeconfig via a flag; in these cases Kilo automatically uses a service account token and the in-cluster ServiceIP for the Kubernetes API. BUT the catch is that this will not work for nodes that are, for example, in a different cloud provider, since the ServiceIP will not be routable. A future proof solution is to always omit mounting the Kubeconfig but to specify a publicly routable endpoint for the API via the --master flag. I mention this in https://github.com/squat/kilo/issues/49#issuecomment-626994195.

If you are running Kilo in a full mesh, then you don't need the internal IPs at all since nothing is using the internal subnet and everything is going through the VPN over the WAN, this should not cause any problems for Kilo

**Comment by eddiewang** [comment]
I did try adding the `master` flag with the ip of my controller node found in the local kube config file. However, this caused certificate issues when kilo attempted to reach out to master.

thanks for the clarification on the internal ips :)

**Comment by squat** [comment]
Hmmm interesting, is the cert presented on the service IP for the API different than the cert presented when accessing the API from outside? If they are different, then indeed this won't work and you'll have to use the k3s.yaml for Kilo (when the nodes need connect from another cloud/region); if the nodes are connecting from within the same private network, then you can use the service account rather than the Kubeconfig and don't need to set the master endpoint, since the internal service IP will be routable. It sounds like you have a good hold on this though :)

**Comment by eddiewang** [comment]
yes, it seems that on most VPS providers (Heztner, Contabo, etc.) with K3S (1.17 or 1.18), the cert must different on the service ip vs accessing from the outside.

I'm trying to figure out what i did to avoid copying the k3s.yaml to the worker nodes.... I think I triggered something special when I passed in a manual token or something, which allowed kilo to work without needing that manual intervention.... which is preferred.

If you have some guidance for me on what could be causing this, I'm happy to take a look and try to tackle the bug.
